Support for Instead of Trigger On Views in Babelfish

deepakshi-mittal opened the pull request:
Currently Postgres does not support Instead of Trigger On Views for DML queries.

On running such a query on VIEW, query rewriting substitutes view with underlying base table.
Added a hook for skipping the substitution for DML query on the VIEW.

This change will add support for Insteed of Triggers for DML queries when using Babelfish.

Task: BABEL-2170

Engine Pr: https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/225

### Test Scenarios Covered ###
- Instead of Trigger (IOT) on View for Insert/Update/Delete queries
- Transition tables inside IOT on View
- IOT Cross database
- IOT Cross Schema
- IOT for Transaction, Commit, Rollback, Rollback to Savepoint
- IOT for Bulk insert
- Same name trigger not allowed for View/Table
- DROP Trigger cross Schema

### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

deepakshi-mittal committed:
Support for Instead of Trigger On Views in Babelfish

Currently Postgres does not support Instead of Trigger
On Views for DML queries.

On running such a query on VIEW, query rewriting substitutes
view with underlying base table.
Added a hook for skipping the substitution for DML
query on the VIEW.

This change will add support for Instaed of Triggers
for DML queries when using Babelfish.

Task: BABEL-2170

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

forestkeeper commented on test/JDBC/expected/BABEL-2170-vu-cleanup.out:
```diff
@@ -0,0 +1,21 @@
+DROP TRIGGER IF EXISTS babel_2170_vu_employees_view_iot_insert;
```
Need to add this test into update schedule too

forestkeeper commented:
Please add more test cases like : 
1. create view trigger for the view cross different babelfish database but with the same view name, verify the right trigger is triggered
2. create view trigger for the view in different babelfish schema but with the same view name, verify the right trigger is triggered
3. Test drop cross db trigger
4. Test the transaction behavior, for example : if one of the stmt failed, the change inside the trigger procedure will all rollbacked or not.

deepakshi-mittal committed:
Added test cases for IOT

- cross db triggers
- cross schema triggers
- Bulk insert and inserted tables

Task: BABEL-2170

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

deepakshi-mittal committed:
Fix failing test cases in github, added to upadte schedule

Task: BABEL-2170

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

deepakshi-mittal committed:
Added transaction commit, save, rollback in IOT

Changed hook name for BBF

Task: BABEL-2170

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

deepakshi-mittal committed:
Removed extra test file and unused drop commands

Task: BABEL-2170

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

deepakshi-mittal committed:
Fix: Moved multi-db mode test cases to end for failing test cases

Task: BABEL-2170

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>

forestkeeper reviewed (dismissed):
LGTM

forestkeeper reviewed (commented):
Would you add a test case for same trigger name on different table in same schema?

deepakshi-mittal committed:
Added cases to check same name trigger for View and table

Added case to check drop trigger in one schema should
not drop same name trigger of different schema

Task: BABEL-2170

Signed-off-by: Deepakshi Mittal <depakshi@amazon.com>
